File: API/models/conversation.py
import json
class Conversation:
    def __init__(self, history=None, new_message="", convo_id=None, user_id=None):
        self.history = history if history else []
        self.new_message = new_message
        self.id =convo_id
        self.user_id = user_id if user_id else None

    def add_turn(self, role, content):
        self.history.append((role, content))

    def set_new_message(self, message):
        self.new_message = message

    def set_user_id(self, user_id):
        self.user_id = user_id
    def __repr__(self):
        return f"Conversation(history={self.history}, new_message='{self.new_message}')"
    def build_conversation_history(self):
        contents = []
        for role, text in self.history:
            contents.append({
                "role": role,
                "parts": [{"text": text}]
            })
        if self.new_message:
            contents.append({
                "role": "user",
                "parts": [{"text": self.new_message}]
            })
        return contents

    def serialize_history(self):
        return json.dumps(self.history)

    @staticmethod
    def deserialize_history(history_json):
        return json.loads(history_json)

File: backend/internal/conversation/conversation.go
package conversation

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
	"unicode"
)

// Turn represents a single exchange in the conversation.
type Turn struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// Conversation captures the state of a chat between a user and the assistant.
type Conversation struct {
	ID         int64
	UserID     int
	History    []Turn
	NewMessage string
	CreatedAt  time.Time
	UpdatedAt  time.Time
}

// New returns a conversation initialised for the supplied user.
func New(userID int) *Conversation {
	return &Conversation{
		UserID:  userID,
		History: make([]Turn, 0),
	}
}

// AddTurn appends a turn to the conversation history.
func (c *Conversation) AddTurn(role, content string) {
	c.History = append(c.History, Turn{
		Role:    role,
		Content: content,
	})
}

// SerializeHistory marshals the conversation history to a JSON string.
func (c *Conversation) SerializeHistory() (string, error) {
	data, err := json.Marshal(c.History)
	if err != nil {
		return "", fmt.Errorf("marshal history: %w", err)
	}
	return string(data), nil
}

// DeserializeHistory converts a JSON string into conversation turns.
func DeserializeHistory(history string) ([]Turn, error) {
	if strings.TrimSpace(history) == "" {
		return []Turn{}, nil
	}
	var turns []Turn
	if err := json.Unmarshal([]byte(history), &turns); err != nil {
		return nil, fmt.Errorf("unmarshal history: %w", err)
	}
	return turns, nil
}

// BuildHistoryPrompt renders the conversation history into a readable prompt segment.
func (c *Conversation) BuildHistoryPrompt() string {
	if len(c.History) == 0 {
		return ""
	}

	var builder strings.Builder
	builder.WriteString("Previous conversation:\n")
	for _, turn := range c.History {
		builder.WriteString(fmt.Sprintf("%s: %s\n", capitaliseRole(turn.Role), turn.Content))
	}
	builder.WriteString("\n")
	return builder.String()
}

func capitaliseRole(role string) string {
	if role == "" {
		return role
	}
	runes := []rune(role)
	runes[0] = unicode.ToUpper(runes[0])
	return string(runes)
}
